node/subsystem/src/util.rs

```diff
@@ -0,0 +1,566 @@
+	const STOP: Self;
+
+	/// If the message variant contains a hash, return it here
+	fn hash(&self) -> Option<Hash>;
+}
+
+/// A JobHandle manages a particular job for a subsystem.
+pub struct JobHandle<ToJob> {
+	abort_handle: future::AbortHandle,
+	to_job: mpsc::Sender<ToJob>,
+	finished: oneshot::Receiver<()>,
+	outgoing_msgs_handle: usize,
+}
+
+impl<ToJob> JobHandle<ToJob> {
+	/// Send a message to the job.
+	pub async fn send_msg(&mut self, msg: ToJob) -> Result<(), Error> {
+		self.to_job.send(msg).await.map_err(Into::into)
+	}
+
+	/// Abort the job without waiting for a graceful shutdown
+	pub fn abort(self) {
+		self.abort_handle.abort();
+	}
+}
+
+impl<ToJob: ToJobTrait> JobHandle<ToJob> {
+	/// Stop this job gracefully.
+	///
+	/// If it hasn't shut itself down after `JOB_GRACEFUL_STOP_DURATION`, abort it.
+	pub async fn stop(mut self) {
+		// we don't actually care if the message couldn't be sent
+		let _ = self.to_job.send(ToJob::STOP).await;
+		let stop_timer = Delay::new(JOB_GRACEFUL_STOP_DURATION);
+
+		match future::select(stop_timer, self.finished).await {
+			Either::Left((_, _)) => {}
+			Either::Right((_, _)) => {
+				self.abort_handle.abort();
+			}
+		}
+	}
+}
+
+/// This trait governs jobs.
+///
+/// Jobs are instantiated and killed automatically on appropriate overseer messages.
+/// Other messages are passed along to and from the job via the overseer to other
+/// subsystems.
+pub trait JobTrait {
+	/// Message type to the job. Typically a subset of AllMessages.
+	type ToJob: 'static + ToJobTrait + Send;
+	/// Message type from the job. Typically a subset of AllMessages.
+	type FromJob: 'static + Into<AllMessages> + Send;
+	/// Job runtime error.
+	type Error: std::fmt::Debug;
+	/// Extra arguments this job needs to run properly.
+	///
+	/// If no extra information is needed, it is perfectly acceptable to set it to `()`.
+	type RunArgs: 'static + Send;
+
+	/// Name of the job, i.e. `CandidateBackingJob`
+	const NAME: &'static str;
+
+	/// Run a job for the parent block indicated
+	fn run(
+		parent: Hash,
+		run_args: Self::RunArgs,
+		rx_to: mpsc::Receiver<Self::ToJob>,
+		tx_from: mpsc::Sender<Self::FromJob>,
+	) -> Pin<Box<dyn Future<Output = Result<(), Self::Error>> + Send>>;
+
+	/// Handle a message which can't be dispatched to a particular job
+	///
+	/// By default, this is implemented with a NOP function. However, if
+	/// ToJob occasionally has messages which do not correspond to a particular
+	/// parent relay hash, then this function will be spawned as a one-off
+	/// task to handle those messages.
+	// TODO: the API here is likely not precisely what we want; figure it out more
+	// once we're implementing a subsystem which actually needs this feature.
+	// In particular, we're quite likely to want this to return a future instead of
+	// interrupting the active thread for the duration of the handler.
+	fn handle_unhashed_msg(_msg: Self::ToJob) -> Result<(), Self::Error> {
+		Ok(())
+	}
```

Comment:
I'm intentionally leaving this underspecified, but we should probably talk about what the API for this case should look like eventually. See also: [usage here](https://github.com/paritytech/polkadot/pull/1376/files#diff-8a0bb46e1e7d7d34cfb3e47e0ce0605aR519-R524)

Reply:
I see, so this is for messages that don't have a relay-parent or no natural relay-parent? I do see the use-case for this, especially for something like the candidate validation subsystem as mentioned above.

bikeshed: I'd not call these "unhashed" messages. unpinned, unanchored, global, idk

Reply:
Yes, exactly. What do you think of "orphan" as a descriptor of a message without a parent?

Reply:
`unanchored` or `parentless` - `orphan` is usually considered equal to `invalid`, so depends on the intention.

Reply:
I'd prefer `unanchored` over `orphan`, since `orphan` has a different connotation in the blockchain space: that it refers to an abandoned fork of the chain